tests/unit/s2n_random_test.c

```diff
@@ -142,6 +146,110 @@ static S2N_RESULT s2n_basic_pattern_tests(S2N_RESULT (*s2n_get_random_data_cb)(s
+    int64_t chosen_upper_bound;
+    struct s2n_blob upper_bound_blob = { .data = (void *) &chosen_upper_bound, .size = sizeof(chosen_upper_bound) };
+
+    /* 0 is not a legal upper bound */
+    current_bound = 0;
+    EXPECT_ERROR(s2n_get_range_cb(current_bound, &current_output));
+
+    /* For an upper bound of 1, 0 should be the only possible output */
+    current_bound = 1;
+    EXPECT_OK(s2n_get_range_cb(current_bound, &current_output));
+    EXPECT_EQUAL(current_output, 0);
+
+    /* For a upper bound of 2, 0 and 1 should be the only possible outputs */
+    current_bound = 1;
+    EXPECT_OK(s2n_get_range_cb(current_bound, &current_output));
+    EXPECT_TRUE((current_output == 0) || (current_output == 1));
+
+    /* Test NUMBER_OF_BOUNDS upper bounds. For each resulting range, draw
+     * NUMBER_OF_RANGE_FUNCTION_CALLS numbers from s2n_public_random() and
+     * verify the output. Set 2^30 * NUMBER_OF_RANGE_FUNCTION_CALLS as the
+     * minimal value for the upper bound.
+     */
+    int64_t minimal_upper_bound = (int64_t) 0x40000000 * (int64_t) NUMBER_OF_RANGE_FUNCTION_CALLS;
+    for (size_t bound_ctr = 0; bound_ctr < NUMBER_OF_BOUNDS; bound_ctr++) {
+
+UPPER_BOUND_TOO_SMALL:
+        EXPECT_OK(s2n_get_private_random_data(&upper_bound_blob));
+        if (chosen_upper_bound < minimal_upper_bound) {
+            goto UPPER_BOUND_TOO_SMALL;
+        }
+
+        for (size_t func_call_ctr = 0; func_call_ctr < NUMBER_OF_RANGE_FUNCTION_CALLS; func_call_ctr++) {
+            EXPECT_OK(s2n_public_random(chosen_upper_bound, &range_results[func_call_ctr]));
+            EXPECT_TRUE(range_results[func_call_ctr] < chosen_upper_bound);
+        }
+
+        /* The probability of "at least MAX_REPEATED_OUTPUT repeated values"
+         * follows a binomial distribution. Hence, we can get an upper bound via
+         * Markov's inequality:
+         * P("at least MAX_REPEATED_OUTPUT repeated values")
+         *      <= E("at least MAX_REPEATED_OUTPUT repeated values") / MAX_REPEATED_OUTPUT.
+         *       = (NUMBER_OF_RANGE_FUNCTION_CALLS * 1/(2^30 * NUMBER_OF_RANGE_FUNCTION_CALLS)) / MAX_REPEATED_OUTPUT
+         *       = 1/(2^30 * MAX_REPEATED_OUTPUT)
+         *
+         * With current parameters
+         *   NUMBER_OF_BOUNDS = 10
+         *   MAX_REPEATED_OUTPUT = 4
+         * this ends up with about a ~1/2^30 probability of failing this test
+         * with a false positive.
+         *
+         * O(n^2) sorting below, but NUMBER_OF_RANGE_FUNCTION_CALLS is very
+         * small, so no big biggie. Sorting the array means that we can check
+         * for reapeated numbers by just counting from left to right resetting
+         * the count when meeting a different value.
+         */
+        sort_array(range_results);
+        uint64_t current_value = range_results[0];
+        size_t repeat_count = 1;
+        for (size_t ctr = 1; ctr < NUMBER_OF_RANGE_FUNCTION_CALLS - 1; ctr++) {
+
```

Comment:
What about a check here to ensure that current value <= range_results[ctr]? Unless I'm reading this wrong(very possible 🙃) that would ensure that your sorting function worked correctly.